### DISCUSSION TITLE
Are there plans to add unit testing / auto tests?

### DISCUSSION THREAD

**Discussion by davehouser1** [foss42/apidash#96] [Ideas & Features]
Having the ability to perform unit tests / auto testing with a suite of requests is a must have. Would be great if your tool included this feature. I strongly recommend including the following sub features into the testing portion if you decide to include this feature:

- Ability to create "unit tests", which is to say a specific request to run, then assert. I think there are modules you can use to allow users to use JavaScript to confirm a test ran properly. You can use frameworks like [Mocha](https://mochajs.org/) framework, and [Chai](https://www.chaijs.com/api/bdd/) for assertions.
- Ability to create test suites, which is an aggregation of unit tests.
- Ability to create test suites of suites, combining suites together.
- Support duplicating unit tests in a suite.
- Support duplicating tests suites.
- Support re-arraigning unit tests in a suite.
- Support for async requests in a given unit test. Just because a request returned a 200OK does not mean the work is complete on the backend. There has to be someway to perform more operations to check if async work is complete. There can be a variety of of ways to assert this. In most cases Iv seen there maybe another address/route that one can check job status of a job. 
- Support for access to environmental variables in tests. Or create a new environmental variable set for a suite or suite of suites. 
- Show how long each unit test took to complete
- Support to show total time all tests in a suite took to run.
- The number of tests can become cumbersome, add a way to search for each unit test in a suite. 
- Ability to "play" or run any unit test along without having to run the whole suite.

Hope to see this one day.

**Comment by animator** [comment]
Hi @davehouser1 

We do want to add the ability to perform unit tests and create test suites in API Dash.
This thread will definitely help in building a roadmap for the same.
It will require work on both design and implementation front so we will have to develop the various features you mentioned (along with other useful testing features) into stages & tackle them.
Thank you for starting this discussion thread and providing your feedback.
Will update it as we progress.

**Comment by davehouser1** [comment]
Adding a couple more features:

- Ability to disable tests in a suite.
- Control if testing should continue or stop immediately if a single test fails. 
- Ability to control async timeout (Chai sets this to a max of 60 seconds, needs to be adjustable)
- Create and export a report. Report includes all tests, pass fail status, time it took, output of failure. Format is up to you, csv, json, whatever you like.

**Comment by animator** [comment]
Thanks for adding more features to the list @davehouser1

**Comment by adityasuhane-06** [comment]
Hi @animator ,

I'm interested in working on this feature as part of my GSoC 2026 proposal for **Idea 4: Agentic API Testing**.

### Proposed Approach: Start with Assertion Framework

I'd like to begin by building the foundation - an **Assertion Framework** for API response validation. This would include:

**Core Features:**
1. **Assertion Types**: Status code, headers, response body (JSON/XML), cookies, response time
2. **Assertion Operators**: equals, notEquals, contains, greaterThan, lessThan, matches (regex), typeOf
3. **JSON Path Support**: Navigate nested properties (e.g., `body.user.profile.email`, `body.items[0].price`)
4. **Data Models**: Store assertions alongside requests in collections
5. **Execution Engine**: Run assertions and generate pass/fail results

**Why This Helps My GSoC Project:**

My GSoC proposal focuses on **AI-powered agentic testing** where AI agents automatically:
- Generate test cases from API specifications
- Execute tests and validate responses
- Self-heal when APIs change
- Provide intelligent feedback on failures

The assertion framework is the **critical foundation** for this because:

1. **AI Test Generator Agent** needs a structured way to create assertions
   - Instead of generating random tests, it creates assertions: "status code should be 200", "response.user.id should exist"
   
2. **Test Validator Agent** needs to evaluate if tests passed/failed
   - The assertion engine provides the validation logic the AI uses
   
3. **Self-Healing Agent** needs to understand what broke
   - When assertions fail, the AI analyzes them to propose fixes
   
4. **Human-in-the-Loop** workflow needs clear test definitions
   - Users review and approve AI-generated assertions before execution

**Example Flow:**
User: "Test my /users API" ↓ AI Generator: Creates assertions automatically

Assert status code equals 200
Assert body.users is array
Assert body.users[0].email matches email regex ↓ Assertion Engine: Executes and validates ↓ AI Validator: Analyzes results, suggests improvements


I've researched industry standards:
- **Postman**: Chai-style assertions with JavaScript
- **REST Assured**: BDD-style (Given-When-Then) with Java
- **Insomnia**: JSON Schema validation
- **API testing best practices**: Status code first, specific property paths, type validation

I can share detailed design docs if helpful.

Looking forward to your feedback!

**Comment by aburiro** [comment]
Hi! I'm Azhar (@aburiro), Flutter developer from Pakistan. I have a merged PR in CCExtractor's taskwarrior-flutter. I'm interested in Idea #4 — Agentic API Testing for GSoC 2026. Issue #1158 seems closed — could a mentor point me to the right issue or discussion for this idea? I'd love to start contributing

**Reply by animator** [reply]
@aburiro https://github.com/foss42/apidash/discussions/1230
